### ISSUE TITLE
ImageFolderDataset

### ISSUE THREAD

**Issue by XishunZhu** [XishunZhu/Garesat_Net-SR-#4]
# encoding: utf-8
import torch.utils.data as data

from PIL import Image
import os
import os.path

from torch.utils.data import DataLoader

import utils.transformed as transforms
IMG_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.ppm', '.bmp', '.pgm']

def is_image_file(filename):
    """Checks if a file is an image.

    Args:
        filename (string): path to a file

    Returns:
        bool: True if the filename ends with a known image extension
    """
    filename_lower = filename.lower()  
    return any(filename_lower.endswith(ext) for ext in IMG_EXTENSIONS)  

def find_classes(dir):

    classes = [d for d in os.listdir(dir) if os.path.isdir(os.path.join(dir, d))]
    classes.sort()  
    class_to_idx = {classes[i]: i for i in range(len(classes))}
    return classes, class_to_idx   

def make_dataset(dir):
    images = []
    labels1 = []
    labels2 = []
    dir = os.path.expanduser(dir)  # Expand the user's home directory
    # for target in sorted(os.listdir(dir)):
    for target in os.listdir(dir):

        if "x8" in target:
            d = os.path.join(dir, target)
            im = Image.open(d)  
            if im.size[0]%4==0 and im.size[1]%4==0:
                # print(im.size[0],im.size[1])
                images.append(d)

                label1 = target.replace("x8", "x4")
                la1 = os.path.join(dir, label1)
                labels1.append(la1)
                label2 = target.replace("x8", "x2")
                la2 = os.path.join(dir, label2)
                labels2.append(la2)

    return images,labels1,labels2

def pil_loader(path):
    # open path as file to avoid ResourceWarning (https://github.com/python-pillow/Pillow/issues/835)
    with open(path, 'rb') as f:  
        with Image.open(f) as img: 
            return img.convert('RGB')

def accimage_loader(path):
    print("can't find acc image loader")
    import accimage
    try:
        return accimage.Image(path)
    except IOError:
        # Potentially a decoding problem, fall back to PIL.Image
        return pil_loader(path)

def default_loader(path):
    from torchvision import get_image_backend
    if get_image_backend() == 'accimage':
        return accimage_loader(path)
    else:
        return pil_loader(path)

class MyImageFolder(data.Dataset):
    """A generic data loader where the images are arranged in this way: ::

        root/dog/xxx.png
        root/dog/xxy.png
        root/dog/xxz.png

        root/cat/123.png
        root/cat/nsdf3.png
        root/cat/asd932_.png

    Args:
        root (string): Root directory path.
        transform (callable, optional): A function/transform that  takes in an PIL image
            and returns a transformed version. E.g, ``transforms.RandomCrop``
        target_transform (callable, optional): A function/transform that takes in the
            target and transforms it.
        loader (callable, optional): A function to load an image given its path.

     Attributes:
        classes (list): List of the class names.
        class_to_idx (dict): Dict with items (class_name, class_index).
        imgs (list): List of (image path, class_index) tuples
    """

    def __init__(self, dir, transform=None):
        global HH
        image=[]
        labels1=[]
        labels2 = []
        for target in os.listdir(dir):
            if "x2" in target:
                d = os.path.join(dir, target)
                im = Image.open(d)  #            
                image.append(d)

                label1 = target.replace("x2", "x4")
                la1 = os.path.join(dir, label1)
                labels1.append(la1)
                label2 = target.replace("x2", "x8")
                la2 = os.path.join(dir, label2)
                labels2.append(la2)
            self.image_list = image
            self.labels1 = labels1
            self.labels2 = labels2
            self.loader = default_loader
            self.transform = transform
            self.len = len(image)

    def __getitem__(self, index):
        """
        Args:
            index (int): Index

        Returns:
            tuple: (image, target) where target is class_index of the target class.
        """
        path = self.image_list[index]

        path1 = self.labels1[index]
        path2 = self.labels2[index]
        img = self.loader(path)
        label1 = self.loader(path1)
        label2 = self.loader(path2)
        if self.transform is not None:
            img = self.transform(img)
            label1 = self.transform(label1)
            label2 = self.transform(label2)
            # print(img.shape)
            # print(label.shape)

        return img,label1,label2

    def __len__(self):
        return self.len
